File: dingtalk.php
#!/usr/bin/env php
<?php

/**
 * 获取和检查URL
 *
 * @param      $arg
 * @param null $default
 * @return mixed
 */
function getArgument($arg , $default = null)
{
    $argv = $GLOBALS['_SERVER']['argv'];
    if (in_array($arg , $argv)) {
        return $argv[array_search($arg , $argv) + 1] ?? $default;
    }
    
    return $default;
}

/**
 * 请求API，获取一言
 *
 * @param $apiUrl
 * @return mixed
 * @copyright ©Alex (System.out.println.org and use TLS). All rights reserved.
 */
function getHitokoto($apiUrl)
{
    return json_decode(file_get_contents($apiUrl));
}

/**
 * 生成数据
 *
 * @param $source
 * @param $creator
 * @param $sentence
 * @return string
 * @copyright ©Alex (System.out.println.org and use TLS). All rights reserved.
 */
function makeData($source , $creator , $sentence)
{
    return json_encode([
        'msgtype'  => 'markdown' ,
        'markdown' => [
            'title' => '一言' ,
            'text'  => "**{$sentence}**
            
> {$creator}「{$source}」" ,
        ] ,
    ]);
}

/**
 * 请求钉钉
 *
 * @param       $uri
 * @param       $data
 * @return mixed
 * @copyright ©Alex (System.out.println.org and use TLS). All rights reserved.
 */
function postToDingtalk($uri , $data)
{
    $ch = curl_init();
    curl_setopt($ch , CURLOPT_URL , $uri);
    curl_setopt($ch , CURLOPT_POST , 1);
    curl_setopt($ch , CURLOPT_CONNECTTIMEOUT , 5);
    curl_setopt($ch , CURLOPT_HTTPHEADER , ['Content-Type: application/json;charset=utf-8']);
    curl_setopt($ch , CURLOPT_POSTFIELDS , $data);
    curl_setopt($ch , CURLOPT_RETURNTRANSFER , true);
    $data = curl_exec($ch);
    curl_close($ch);
    
    return json_decode($data);
}

// 钉钉机器人 URL
$robotUrl = getArgument('--robot-url');
if (! preg_match('/^https?:\/\//' , $robotUrl)) {
    exit('--robot-url must be url');
}

// 一言API地址
$apiUrl = getArgument('--api-url' , 'https://v1.hitokoto.cn/#');
if (! preg_match('/^https?:\/\//' , $apiUrl)) {
    exit('--api-url must be url');
}

// 请求
$hitokoto = getHitokoto($apiUrl);
// 构造 markdown 和发送到钉钉
$result = postToDingtalk(
    $robotUrl ,
    makeData($hitokoto->from , $hitokoto->creator , $hitokoto->hitokoto)
);

// 判断响应
if ($result->errcode !== 0) {
    exit($result->errmsg . PHP_EOL);
}

exit('success');

File: dingtalk.go
package main

import (
    "bytes"
    "encoding/json"
    "fmt"
    "io"
    "io/ioutil"
    "net/http"
    "os"
    "regexp"
)

// 获取一言
func getHitokoto(url string) map[string]interface{} {
    response, error := http.Get(url)
    if error != nil {
        panic(error)
    }

    defer response.Body.Close()
    return json2Map(response.Body)
}

// 从命令行获取参数
func getArgument(name string, defaultValue string) string {
    args := os.Args[1:]
    length := len(args)
    for index, value := range args {
        if value == name {
            if index+1 <= length {
                return args[index+1]
            }
        }
    }

    return defaultValue
}

// 获取参数并设置默认值为空字符串
func getArgDefaultEmpty(name string) string {
    return getArgument(name, "")
}

// 将json数据从字节数组转换成map
func json2Map(jsonString io.Reader) map[string]interface{} {
    body, err := ioutil.ReadAll(jsonString)
    if err != nil {
        panic(err)
    }

    // 解析 json
    result := make(map[string]interface{})
    err = json.Unmarshal(body, &result)
    if err != nil {
        panic(err)
    }

    return result
}

// 发送json请求，向钉钉推数据
// 已知 url 和 json 都是合法的，忽略其它异常
func postWithJson(robotUrl string, data []byte) map[string]interface{} {
    request, _ := http.NewRequest("POST", robotUrl, bytes.NewBuffer(data))
    request.Header.Set("Content-Type", "application/json")

    client := &http.Client{}
    response, err := client.Do(request)
    if err != nil {
        panic(err)
    }

    defer response.Body.Close()

    return json2Map(response.Body)
}

// 构造推送向钉钉的数据
// 数据结构是清晰的，数据变量也是已知的，所以忽略错误
func makeData(source string, creator string, sentence string) []byte {
    jsonData, _ := json.Marshal(map[string]interface{}{
        "msgtype": "markdown",
        "markdown": map[string]string{
            "title": "一言",
            "text":  "**" + sentence + "** \n > " + creator + "「" + source + "」",
        },
    })

    return jsonData
}

func main() {
    // 构造正则，验证输入参数是否是合法的 url
    // 正则固定，忽略异常
    regs, _ := regexp.Compile("^https?://")

    apiUrl := getArgDefaultEmpty("--api-url")
    if !regs.MatchString(apiUrl) {
        fmt.Println("--api-url is required")
        os.Exit(127)
    }

    robotUrl := getArgDefaultEmpty("--robot-url")
    if !regs.MatchString(robotUrl) {
        fmt.Println("--robot-url must be url")
        os.Exit(125)
    }

    // 获取数据和推送到钉钉
    hitokoto := getHitokoto(apiUrl)
    response := postWithJson(robotUrl, makeData(
        hitokoto["from"].(string),
        hitokoto["creator"].(string),
        hitokoto["hitokoto"].(string),
    ))

    errCode := int(response["errcode"].(float64))
    if errCode != 0 {
        fmt.Println(response["errmsg"])
        os.Exit(errCode)
    }

    fmt.Println("success")
}
